core/vm/runtime/consensus_fault.hpp

```diff
@@ -0,0 +1,136 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include "cbor_blake/ipld_version.hpp"
+#include "primitives/block/rand.hpp"
+#include "primitives/tipset/chain.hpp"
+#include "vm/actor/builtin/states/miner/miner_actor_state.hpp"
+#include "vm/actor/builtin/types/miner/policy.hpp"
+#include "vm/runtime/env_context.hpp"
+#include "vm/runtime/pricelist.hpp"
+#include "vm/state/impl/state_tree_impl.hpp"
+#include "vm/state/resolve_key.hpp"
+
+namespace fc::vm::runtime {
+  using actor::builtin::states::MinerActorStatePtr;
+  using actor::builtin::types::miner::kChainFinality;
+  using primitives::block::BlockHeader;
+  using state::StateTreeImpl;
+
+  template <typename T>
+  inline bool has(const std::vector<T> &xs, const T &x) {
+    return std::find(xs.begin(), xs.end(), x) != xs.end();
+  }
+
+  inline bool isNearOrange(ChainEpoch epoch) {
+    return epoch > kUpgradeOrangeHeight - kChainFinality
+           && epoch < kUpgradeOrangeHeight + kChainFinality;
+  }
+
+  struct ReadGasIpld : Ipld {
+    const IpldPtr &ipld;
+    const Pricelist &pricelist;
+    mutable GasAmount gas{};
+
+    ReadGasIpld(const IpldPtr &ipld, const Pricelist &pricelist)
+        : ipld{ipld}, pricelist{pricelist} {}
+
+    outcome::result<bool> contains(const CID &key) const override {
+      return ERROR_TEXT("ReadGasIpld.contains");
+    }
+    outcome::result<void> set(const CID &key, BytesCow &&value) override {
+      return ERROR_TEXT("ReadGasIpld.set");
+    }
+    outcome::result<Value> get(const CID &key) const override {
+      gas += pricelist.onIpldGet();
+      return ipld->get(key);
+    }
+  };
+
+  inline outcome::result<ConsensusFault> consensusFault(
+      GasAmount &gas,
+      const EnvironmentContext &envx,
+      const TsBranchPtr &ts_branch,
+      ChainEpoch epoch,
+      const CID &base_state,
+      BytesIn block1,
+      BytesIn block2,
+      BytesIn extra) {
+    const auto kNoFault{ERROR_TEXT("consensusFault")};
```

Comment:
That's lot of code here in `.hpp`. Please, consider to move to `.cpp`.
It would be great if you add a short comment what the function does.

Reply:
Won't move, it's called in only 2 places.